app/models/host.rb

```diff
@@ -47,6 +47,8 @@ class Host < ApplicationRecord
   has_many                  :miq_templates, :inverse_of => :host
   has_many                  :host_storages, :dependent => :destroy
   has_many                  :storages, :through => :host_storages
+  has_many                  :writable_accessible_host_storages, -> { writable_accessible }, :class_name => "HostStorage"
+  has_many                  :writable_accessible_storages, :through => :writable_accessible_host_storages, :source => :storage
```

Comment:
I think it would be good to replace the [`writable_storages`](https://github.com/ManageIQ/manageiq/blob/afddd9479639d504bd157260e7377359b9c75bd7/app/models/host.rb#L1739-L1745) and [`read_only_storages`](https://github.com/ManageIQ/manageiq/blob/afddd9479639d504bd157260e7377359b9c75bd7/app/models/host.rb#L1747-L1753) with associations using these scopes, not sure if in this PR or a followup since it isn't really related

Reply:
/cc @NickLaMuro since you modified the `writable_storages` method for performance, wdyt?

Reply:
For reference (since I had to go find it), that performance change was done here:

https://github.com/ManageIQ/manageiq/pull/17354

Reply:
Okay, so I think my to cents with this is that you probably want to leave those methods (`writable_storages` and `read_only_storages`), and just change them to use the scopes:

```diff
 def writable_storages
    if host_storages.loaded? && host_storages.all? { |hs| hs.association(:storage).loaded? }
       host_storages.reject(&:read_only).map(&:storage)
     else
-      storages.where(:host_storages => {:read_only => [false, nil]})
+      storages.writable  # I think my pseudo code here is wrong...
    end
  end
```

Or, if we did create a `writable_storages` scope, I would suggest possibly doing the following to maintain the performance benefits of https://github.com/ManageIQ/manageiq/pull/17354 :


```ruby
has_many                  :writable_host_storages, -> { writable }, :class_name => "HostStorage"
has_many                  :writable_storages, :through => :writable_host_storages, :source => :storage

# overwrite default reader method
def writable_storages
  if association(:writable_storages).loaded? # use already loaded the relationship (faster)
    association(:writable_storages).reader
  # can use an existing relationship to avoid another query and instantiation of other records (requires loop)
  elsif host_storages.loaded? && host_storages.all? { |hs| hs.association(:storage).loaded? }
    host_storages.reject(&:read_only).map(&:storage)
  else # call the relationship
    association(:writable_storages).reader
  end
end
```

Let me know if any of this needs a further explanation... trying to avoid a :book:

Reply:
Alright I'll leave it as is :smile: